packages/@uppy/aws-s3/src/MiniXHRUpload.js

```diff
@@ -0,0 +1,360 @@
+  constructor (uppy, opts) {
+    this.uppy = uppy
+    this.opts = {
+      validateStatus (status, responseText, response) {
+        return status >= 200 && status < 300
+      },
+      ...opts
+    }
+
+    this.requests = opts.__queue
+    this.uploaderEvents = Object.create(null)
+  }
+
+  _getOptions (file) {
+    const uppy = this.uppy
+
+    const overrides = uppy.getState().xhrUpload
+    const opts = {
+      ...this.opts,
+      ...(overrides || {}),
+      ...(file.xhrUpload || {}),
+      headers: {}
+    }
+    Object.assign(opts.headers, this.opts.headers)
+    if (overrides) {
+      Object.assign(opts.headers, overrides.headers)
+    }
+    if (file.xhrUpload) {
+      Object.assign(opts.headers, file.xhrUpload.headers)
+    }
+
+    return opts
+  }
+
+  uploadFile (id, current, total) {
+    const file = this.uppy.getFile(id)
+    if (file.error) {
+      throw new Error(file.error)
+    } else if (file.isRemote) {
+      return this._uploadRemoteFile(file, current, total)
+    }
+    return this._uploadLocalFile(file, current, total)
+  }
+
+  _addMetadata (formData, meta, opts) {
+    const metaFields = Array.isArray(opts.metaFields)
+      ? opts.metaFields
+      // Send along all fields by default.
+      : Object.keys(meta)
+    metaFields.forEach((item) => {
+      formData.append(item, meta[item])
+    })
+  }
+
+  _createFormDataUpload (file, opts) {
+    const formPost = new FormData()
+
+    this._addMetadata(formPost, file.meta, opts)
+
+    const dataWithUpdatedType = setTypeInBlob(file)
+
+    if (file.name) {
+      formPost.append(opts.fieldName, dataWithUpdatedType, file.meta.name)
+    } else {
+      formPost.append(opts.fieldName, dataWithUpdatedType)
+    }
+
+    return formPost
+  }
+
+  _createBareUpload (file, opts) {
+    return file.data
+  }
+
+  _onFileRemoved (fileID, cb) {
+    this.uploaderEvents[fileID].on('file-removed', (file) => {
+      if (fileID === file.id) cb(file.id)
+    })
+  }
+
+  _onRetry (fileID, cb) {
+    this.uploaderEvents[fileID].on('upload-retry', (targetFileID) => {
+      if (fileID === targetFileID) {
+        cb()
+      }
+    })
+  }
+
+  _onRetryAll (fileID, cb) {
+    this.uploaderEvents[fileID].on('retry-all', (filesToRetry) => {
+      if (!this.uppy.getFile(fileID)) return
+      cb()
+    })
+  }
+
+  _onCancelAll (fileID, cb) {
+    this.uploaderEvents[fileID].on('cancel-all', () => {
+      if (!this.uppy.getFile(fileID)) return
+      cb()
+    })
+  }
+
+  _uploadLocalFile (file, current, total) {
+    const opts = this._getOptions(file)
+
+    this.uppy.log(`uploading ${current} of ${total}`)
+    return new Promise((resolve, reject) => {
+      // This is done in index.js in the S3 plugin.
+      // this.uppy.emit('upload-started', file)
+
+      const data = opts.formData
+        ? this._createFormDataUpload(file, opts)
+        : this._createBareUpload(file, opts)
+
+      const timer = new ProgressTimeout(opts.timeout, () => {
+        xhr.abort()
```

Comment:
I got a lil confused here cos I couldn't find anywhere `xhr` was defined before it was referenced. Would it make sense to move [this](https://github.com/transloadit/uppy/pull/2060/files#diff-57355ea8780221946e05753c979ce54eR152) right before `ProgressTimeout` is instantiated?

Reply:
hmm yes, the ordering is going to be a little confusing no matter what (`queuedRequest` has the same problem), but for this `timer` variable we can solve it by moving the xhr stuff up :+1: